lib/pos_connection_controller.go

```diff
@@ -0,0 +1,722 @@
+	// it's aware of at random and provides it to us.
+	AddrMgr *addrmgr.AddrManager
+
+	// addrsToBroadcast is a list of all the addresses we've received from valid addr
+	// messages that we intend to broadcast to our peers. It is organized as:
+	// <recipient address> -> <list of addresses we received from that recipient>.
+	//
+	// It is organized in this way so that we can limit the number of addresses we
+	// are distributing for a single peer to avoid a DOS attack.
+	addrsToBroadcastLock deadlock.RWMutex
+	addrsToBroadcast     map[string][]*SingleAddr
+
+	// When --connectips is set, we don't connect to anything from the addrmgr.
+	connectIps []string
+
+	// The target number of outbound peers we want to have.
+	targetOutboundPeers uint32
+	// The maximum number of inbound peers we allow.
+	maxInboundPeers uint32
+	// When true, only one connection per IP is allowed. Prevents eclipse attacks
+	// among other things.
+	limitOneInboundConnectionPerIP bool
+
+	startGroup sync.WaitGroup
+	exitChan   chan struct{}
+	exitGroup  sync.WaitGroup
+}
+
+func NewConnectionController(params *DeSoParams, cmgr *ConnectionManager, rnManager *RemoteNodeManager, signer *BLSSigner,
+	addrMgr *addrmgr.AddrManager, targetOutboundPeers uint32, maxInboundPeers uint32,
+	limitOneInboundConnectionPerIP bool) *ConnectionController {
+
+	return &ConnectionController{
+		params:                         params,
+		cmgr:                           cmgr,
+		signer:                         signer,
+		rnManager:                      rnManager,
+		AddrMgr:                        addrMgr,
+		addrsToBroadcast:               make(map[string][]*SingleAddr),
+		targetOutboundPeers:            targetOutboundPeers,
+		maxInboundPeers:                maxInboundPeers,
+		limitOneInboundConnectionPerIP: limitOneInboundConnectionPerIP,
+		exitChan:                       make(chan struct{}),
+	}
+}
+
+func (cc *ConnectionController) Start() {
+	cc.startGroup.Add(3)
+	// Start the validator connector
+	go cc.startValidatorConnector()
+
+	cc.startGroup.Wait()
+	cc.exitGroup.Add(3)
+}
+
+func (cc *ConnectionController) Stop() {
+	close(cc.exitChan)
+	cc.exitGroup.Wait()
+}
+
+func (cc *ConnectionController) GetRemoteNodeManager() *RemoteNodeManager {
+	return cc.rnManager
+}
+
+func (cc *ConnectionController) initiatePersistentConnections() {
+	// This is a hack to make outbound connections go away.
+	if cc.targetOutboundPeers == 0 {
+		return
+	}
+	if len(cc.connectIps) > 0 {
+		// Connect to addresses passed via the --connect-ips flag. These addresses
+		// are persistent in the sense that if we disconnect from one, we will
+		// try to reconnect to the same one.
+		for _, connectIp := range cc.connectIps {
+			if err := cc.createPersistentOutboundConnection(connectIp); err != nil {
+				glog.Errorf("ConnectionController.initiatePersistentConnections: Problem connecting "+
+					"to connectIp %v: %v", connectIp, err)
+			}
+		}
+	}
```

Comment:
Reduce nesting to lower cognitive complexity.
```suggestion
	if len(cc.connectIps) == 0 {
		return
	}
	// Connect to addresses passed via the --connect-ips flag. These addresses
	// are persistent in the sense that if we disconnect from one, we will
	// try to reconnect to the same one.
	for _, connectIp := range cc.connectIps {
		if err := cc.createPersistentOutboundConnection(connectIp); err != nil {
			glog.Errorf("ConnectionController.initiatePersistentConnections: Problem connecting "+
				"to connectIp %v: %v", connectIp, err)
		}
	}
```